File: 1021. Remove Outermost Parentheses/Solution.py
class Solution:
    def removeOuterParentheses(self, S: str) -> str:
        sRes = ""
        
        sOpen = False
        counter = 0
        
        for ch in S:
            if ch == '(':
                if not sOpen:
                    sOpen = True
                    continue
                else:
                    counter += 1
                    sRes += '('
                
            if ch == ')':
                if sOpen and counter == 0:
                    sOpen = False
                    continue
                else:
                    sRes += ')'
                    counter -= 1
                
                
        return sRes
                

File: 1021. Remove Outermost Parentheses/Solution.cpp
class Solution {
public:
    string removeOuterParentheses(string S) {
        std::string sRes;
        sRes.reserve(S.size());
        int iCounter = 0;
        bool bOpen = false;
        
        for (const char &ch : S)
        {
            if (ch == '(')
            {
                if (!bOpen)
                {
                    bOpen = true;
                    continue;
                }
                else
                {
                    iCounter += 1;
                    sRes += '(';
                }
            }
            else
            {
                if (bOpen && iCounter == 0)
                {
                    bOpen = false;
                    continue;
                }
                else
                {
                    iCounter -= 1;
                    sRes += ')';
                }
            }
            
        }
        
        return sRes;
        
    }
};
